Sphinx extension for adding Remix links to code snippets

cameel opened the pull request:
Fixes #11415.

The PR adds a local Sphinx extension that encodes snippets into links and puts a small Remix icon in the top-right corner of every box with highlighted Solidity code.

A few remarks:
- I think it looks decent and works in both dark and light mode. If anyone wants to try and make it look more appealing though, feel free to push more commits to the PR.
- I tried to add a `title` attribute to the link but eventually gave up so the link has no description if you hover over it. I'd have to dig deeper into docutils to find out how to do it (if it's supported at all).
    - I could also use an `<img>` tag instead of CSS background and add `alt` text. I did not want to do it that way because semantically the image is just pure style here. I think it's better if restyling does not require touching Python code.
- Yul snippets get skipped. Not sure if Remix supports those.

Here's how it looks like now:

![solidity-snippet-with-remix-link-light-mode](https://user-images.githubusercontent.com/137030/127036512-92dd8e45-3183-452b-9415-246c9df7657a.png)
![solidity-snippet-with-remix-link-dark-mode](https://user-images.githubusercontent.com/137030/127036506-9993e134-29be-4528-983f-f67aa54a074a.png)

cameel committed:
Sphinx extension for adding Remix links to code snippets

cameel commented on docs/_static/css/custom.css:
```diff
@@ -12,3 +12,31 @@ pre {
+/* Link to Remix IDE shown over code snippets */
+a.remix-link {
+    /* Remix icon. */
+    background:url("https://remix.ethereum.org/assets/img/remixLogo.webp");
```
The icon is served directly from https://remix.ethereum.org. I could instead copy it and put it in the repo. It's a small (5 kB) .webp file. Not sure which is better. Either way, we should probably ask them if hotlinking/copying it is fine.

We might also want to swap it out for something else.

axic replied:
I'd keep a local copy. Remote URLs can change.

cameel replied:
Added the image to the PR. We'll probably replace it with a different one though.

axic commented on docs/ext/remix_code_links.py:
```diff
@@ -0,0 +1,53 @@
+    # NOTE: base64 encoded data may contain +, = and / characters. Remix seems to handle them just
+    # fine without any escaping.
+    base64_encoded_source = base64.b64encode(source_code.encode('utf-8')).decode('ascii')
+    return f"https://remix.ethereum.org/?code={base64_encoded_source}"
```
https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers

axic replied:
Does remix has a version parameter still, or will it just find it out based on the pragma?

axic replied:
Does remix has a version parameter still, or will it just find it out based on the pragma?

cameel replied:
That's a good point. Still, I'm not sure if we can do anything reasonable about that. I could make it not render the link if it's too long but the limit is not consistent across browsers so I think it's better to keep it for those browsers that do support longer URLs.

We could ask Remix to provide a `POST` endpoint that would allow posting longer code.

We could also use a URL shortener at build time but that sounds complicated and brittle.

axic replied:
> We could ask Remix to provide a POST endpoint that would allow posting longer code.

Can do that easily from the user side.

> We could also use a URL shortener at build time but that sounds complicated and brittle.

Just ask remix to support lib compression and/or ascii85 (with a modified characters removing `%` and `&`) in the payload :)

axic replied:
At the minimum you can output a warning on `make docs` if any snippet has an URL longer than 2000 bytes. We could then see if anything is affected.

cameel replied:
The "Voting" example seems to be the longest Solidity snippet currently. It's a bit under 8 kB. [The link that gets generated for it](https://remix.ethereum.org/?code=Ly8gU1BEWC1MaWNlbnNlLUlkZW50aWZpZXI6IEdQTC0zLjAKcHJhZ21hIHNvbGlkaXR5ID49MC43LjAgPDAuOS4wOwovLy8gQHRpdGxlIFZvdGluZyB3aXRoIGRlbGVnYXRpb24uCmNvbnRyYWN0IEJhbGxvdCB7CiAgICAvLyBUaGlzIGRlY2xhcmVzIGEgbmV3IGNvbXBsZXggdHlwZSB3aGljaCB3aWxsCiAgICAvLyBiZSB1c2VkIGZvciB2YXJpYWJsZXMgbGF0ZXIuCiAgICAvLyBJdCB3aWxsIHJlcHJlc2VudCBhIHNpbmdsZSB2b3Rlci4KICAgIHN0cnVjdCBWb3RlciB7CiAgICAgICAgdWludCB3ZWlnaHQ7IC8vIHdlaWdodCBpcyBhY2N1bXVsYXRlZCBieSBkZWxlZ2F0aW9uCiAgICAgICAgYm9vbCB2b3RlZDsgIC8vIGlmIHRydWUsIHRoYXQgcGVyc29uIGFscmVhZHkgdm90ZWQKICAgICAgICBhZGRyZXNzIGRlbGVnYXRlOyAvLyBwZXJzb24gZGVsZWdhdGVkIHRvCiAgICAgICAgdWludCB2b3RlOyAgIC8vIGluZGV4IG9mIHRoZSB2b3RlZCBwcm9wb3NhbAogICAgfQoKICAgIC8vIFRoaXMgaXMgYSB0eXBlIGZvciBhIHNpbmdsZSBwcm9wb3NhbC4KICAgIHN0cnVjdCBQcm9wb3NhbCB7CiAgICAgICAgYnl0ZXMzMiBuYW1lOyAgIC8vIHNob3J0IG5hbWUgKHVwIHRvIDMyIGJ5dGVzKQogICAgICAgIHVpbnQgdm90ZUNvdW50OyAvLyBudW1iZXIgb2YgYWNjdW11bGF0ZWQgdm90ZXMKICAgIH0KCiAgICBhZGRyZXNzIHB1YmxpYyBjaGFpcnBlcnNvbjsKCiAgICAvLyBUaGlzIGRlY2xhcmVzIGEgc3RhdGUgdmFyaWFibGUgdGhhdAogICAgLy8gc3RvcmVzIGEgYFZvdGVyYCBzdHJ1Y3QgZm9yIGVhY2ggcG9zc2libGUgYWRkcmVzcy4KICAgIG1hcHBpbmcoYWRkcmVzcyA9PiBWb3RlcikgcHVibGljIHZvdGVyczsKCiAgICAvLyBBIGR5bmFtaWNhbGx5LXNpemVkIGFycmF5IG9mIGBQcm9wb3NhbGAgc3RydWN0cy4KICAgIFByb3Bvc2FsW10gcHVibGljIHByb3Bvc2FsczsKCiAgICAvLy8gQ3JlYXRlIGEgbmV3IGJhbGxvdCB0byBjaG9vc2Ugb25lIG9mIGBwcm9wb3NhbE5hbWVzYC4KICAgIGNvbnN0cnVjdG9yKGJ5dGVzMzJbXSBtZW1vcnkgcHJvcG9zYWxOYW1lcykgewogICAgICAgIGNoYWlycGVyc29uID0gbXNnLnNlbmRlcjsKICAgICAgICB2b3RlcnNbY2hhaXJwZXJzb25dLndlaWdodCA9IDE7CgogICAgICAgIC8vIEZvciBlYWNoIG9mIHRoZSBwcm92aWRlZCBwcm9wb3NhbCBuYW1lcywKICAgICAgICAvLyBjcmVhdGUgYSBuZXcgcHJvcG9zYWwgb2JqZWN0IGFuZCBhZGQgaXQKICAgICAgICAvLyB0byB0aGUgZW5kIG9mIHRoZSBhcnJheS4KICAgICAgICBmb3IgKHVpbnQgaSA9IDA7IGkgPCBwcm9wb3NhbE5hbWVzLmxlbmd0aDsgaSsrKSB7CiAgICAgICAgICAgIC8vIGBQcm9wb3NhbCh7Li4ufSlgIGNyZWF0ZXMgYSB0ZW1wb3JhcnkKICAgICAgICAgICAgLy8gUHJvcG9zYWwgb2JqZWN0IGFuZCBgcHJvcG9zYWxzLnB1c2goLi4uKWAKICAgICAgICAgICAgLy8gYXBwZW5kcyBpdCB0byB0aGUgZW5kIG9mIGBwcm9wb3NhbHNgLgogICAgICAgICAgICBwcm9wb3NhbHMucHVzaChQcm9wb3NhbCh7CiAgICAgICAgICAgICAgICBuYW1lOiBwcm9wb3NhbE5hbWVzW2ldLAogICAgICAgICAgICAgICAgdm90ZUNvdW50OiAwCiAgICAgICAgICAgIH0pKTsKICAgICAgICB9CiAgICB9CgogICAgLy8gR2l2ZSBgdm90ZXJgIHRoZSByaWdodCB0byB2b3RlIG9uIHRoaXMgYmFsbG90LgogICAgLy8gTWF5IG9ubHkgYmUgY2FsbGVkIGJ5IGBjaGFpcnBlcnNvbmAuCiAgICBmdW5jdGlvbiBnaXZlUmlnaHRUb1ZvdGUoYWRkcmVzcyB2b3RlcikgcHVibGljIHsKICAgICAgICAvLyBJZiB0aGUgZmlyc3QgYXJndW1lbnQgb2YgYHJlcXVpcmVgIGV2YWx1YXRlcwogICAgICAgIC8vIHRvIGBmYWxzZWAsIGV4ZWN1dGlvbiB0ZXJtaW5hdGVzIGFuZCBhbGwKICAgICAgICAvLyBjaGFuZ2VzIHRvIHRoZSBzdGF0ZSBhbmQgdG8gRXRoZXIgYmFsYW5jZXMKICAgICAgICAvLyBhcmUgcmV2ZXJ0ZWQuCiAgICAgICAgLy8gVGhpcyB1c2VkIHRvIGNvbnN1bWUgYWxsIGdhcyBpbiBvbGQgRVZNIHZlcnNpb25zLCBidXQKICAgICAgICAvLyBub3QgYW55bW9yZS4KICAgICAgICAvLyBJdCBpcyBvZnRlbiBhIGdvb2QgaWRlYSB0byB1c2UgYHJlcXVpcmVgIHRvIGNoZWNrIGlmCiAgICAgICAgLy8gZnVuY3Rpb25zIGFyZSBjYWxsZWQgY29ycmVjdGx5LgogICAgICAgIC8vIEFzIGEgc2Vjb25kIGFyZ3VtZW50LCB5b3UgY2FuIGFsc28gcHJvdmlkZSBhbgogICAgICAgIC8vIGV4cGxhbmF0aW9uIGFib3V0IHdoYXQgd2VudCB3cm9uZy4KICAgICAgICByZXF1aXJlKAogICAgICAgICAgICBtc2cuc2VuZGVyID09IGNoYWlycGVyc29uLAogICAgICAgICAgICAiT25seSBjaGFpcnBlcnNvbiBjYW4gZ2l2ZSByaWdodCB0byB2b3RlLiIKICAgICAgICApOwogICAgICAgIHJlcXVpcmUoCiAgICAgICAgICAgICF2b3RlcnNbdm90ZXJdLnZvdGVkLAogICAgICAgICAgICAiVGhlIHZvdGVyIGFscmVhZHkgdm90ZWQuIgogICAgICAgICk7CiAgICAgICAgcmVxdWlyZSh2b3RlcnNbdm90ZXJdLndlaWdodCA9PSAwKTsKICAgICAgICB2b3RlcnNbdm90ZXJdLndlaWdodCA9IDE7CiAgICB9CgogICAgLy8vIERlbGVnYXRlIHlvdXIgdm90ZSB0byB0aGUgdm90ZXIgYHRvYC4KICAgIGZ1bmN0aW9uIGRlbGVnYXRlKGFkZHJlc3MgdG8pIHB1YmxpYyB7CiAgICAgICAgLy8gYXNzaWducyByZWZlcmVuY2UKICAgICAgICBWb3RlciBzdG9yYWdlIHNlbmRlciA9IHZvdGVyc1ttc2cuc2VuZGVyXTsKICAgICAgICByZXF1aXJlKCFzZW5kZXIudm90ZWQsICJZb3UgYWxyZWFkeSB2b3RlZC4iKTsKCiAgICAgICAgcmVxdWlyZSh0byAhPSBtc2cuc2VuZGVyLCAiU2VsZi1kZWxlZ2F0aW9uIGlzIGRpc2FsbG93ZWQuIik7CgogICAgICAgIC8vIEZvcndhcmQgdGhlIGRlbGVnYXRpb24gYXMgbG9uZyBhcwogICAgICAgIC8vIGB0b2AgYWxzbyBkZWxlZ2F0ZWQuCiAgICAgICAgLy8gSW4gZ2VuZXJhbCwgc3VjaCBsb29wcyBhcmUgdmVyeSBkYW5nZXJvdXMsCiAgICAgICAgLy8gYmVjYXVzZSBpZiB0aGV5IHJ1biB0b28gbG9uZywgdGhleSBtaWdodAogICAgICAgIC8vIG5lZWQgbW9yZSBnYXMgdGhhbiBpcyBhdmFpbGFibGUgaW4gYSBibG9jay4KICAgICAgICAvLyBJbiB0aGlzIGNhc2UsIHRoZSBkZWxlZ2F0aW9uIHdpbGwgbm90IGJlIGV4ZWN1dGVkLAogICAgICAgIC8vIGJ1dCBpbiBvdGhlciBzaXR1YXRpb25zLCBzdWNoIGxvb3BzIG1pZ2h0CiAgICAgICAgLy8gY2F1c2UgYSBjb250cmFjdCB0byBnZXQgInN0dWNrIiBjb21wbGV0ZWx5LgogICAgICAgIHdoaWxlICh2b3RlcnNbdG9dLmRlbGVnYXRlICE9IGFkZHJlc3MoMCkpIHsKICAgICAgICAgICAgdG8gPSB2b3RlcnNbdG9dLmRlbGVnYXRlOwoKICAgICAgICAgICAgLy8gV2UgZm91bmQgYSBsb29wIGluIHRoZSBkZWxlZ2F0aW9uLCBub3QgYWxsb3dlZC4KICAgICAgICAgICAgcmVxdWlyZSh0byAhPSBtc2cuc2VuZGVyLCAiRm91bmQgbG9vcCBpbiBkZWxlZ2F0aW9uLiIpOwogICAgICAgIH0KCiAgICAgICAgLy8gU2luY2UgYHNlbmRlcmAgaXMgYSByZWZlcmVuY2UsIHRoaXMKICAgICAgICAvLyBtb2RpZmllcyBgdm90ZXJzW21zZy5zZW5kZXJdLnZvdGVkYAogICAgICAgIHNlbmRlci52b3RlZCA9IHRydWU7CiAgICAgICAgc2VuZGVyLmRlbGVnYXRlID0gdG87CiAgICAgICAgVm90ZXIgc3RvcmFnZSBkZWxlZ2F0ZV8gPSB2b3RlcnNbdG9dOwogICAgICAgIGlmIChkZWxlZ2F0ZV8udm90ZWQpIHsKICAgICAgICAgICAgLy8gSWYgdGhlIGRlbGVnYXRlIGFscmVhZHkgdm90ZWQsCiAgICAgICAgICAgIC8vIGRpcmVjdGx5IGFkZCB0byB0aGUgbnVtYmVyIG9mIHZvdGVzCiAgICAgICAgICAgIHByb3Bvc2Fsc1tkZWxlZ2F0ZV8udm90ZV0udm90ZUNvdW50ICs9IHNlbmRlci53ZWlnaHQ7CiAgICAgICAgfSBlbHNlIHsKICAgICAgICAgICAgLy8gSWYgdGhlIGRlbGVnYXRlIGRpZCBub3Qgdm90ZSB5ZXQsCiAgICAgICAgICAgIC8vIGFkZCB0byBoZXIgd2VpZ2h0LgogICAgICAgICAgICBkZWxlZ2F0ZV8ud2VpZ2h0ICs9IHNlbmRlci53ZWlnaHQ7CiAgICAgICAgfQogICAgfQoKICAgIC8vLyBHaXZlIHlvdXIgdm90ZSAoaW5jbHVkaW5nIHZvdGVzIGRlbGVnYXRlZCB0byB5b3UpCiAgICAvLy8gdG8gcHJvcG9zYWwgYHByb3Bvc2Fsc1twcm9wb3NhbF0ubmFtZWAuCiAgICBmdW5jdGlvbiB2b3RlKHVpbnQgcHJvcG9zYWwpIHB1YmxpYyB7CiAgICAgICAgVm90ZXIgc3RvcmFnZSBzZW5kZXIgPSB2b3RlcnNbbXNnLnNlbmRlcl07CiAgICAgICAgcmVxdWlyZShzZW5kZXIud2VpZ2h0ICE9IDAsICJIYXMgbm8gcmlnaHQgdG8gdm90ZSIpOwogICAgICAgIHJlcXVpcmUoIXNlbmRlci52b3RlZCwgIkFscmVhZHkgdm90ZWQuIik7CiAgICAgICAgc2VuZGVyLnZvdGVkID0gdHJ1ZTsKICAgICAgICBzZW5kZXIudm90ZSA9IHByb3Bvc2FsOwoKICAgICAgICAvLyBJZiBgcHJvcG9zYWxgIGlzIG91dCBvZiB0aGUgcmFuZ2Ugb2YgdGhlIGFycmF5LAogICAgICAgIC8vIHRoaXMgd2lsbCB0aHJvdyBhdXRvbWF0aWNhbGx5IGFuZCByZXZlcnQgYWxsCiAgICAgICAgLy8gY2hhbmdlcy4KICAgICAgICBwcm9wb3NhbHNbcHJvcG9zYWxdLnZvdGVDb3VudCArPSBzZW5kZXIud2VpZ2h0OwogICAgfQoKICAgIC8vLyBAZGV2IENvbXB1dGVzIHRoZSB3aW5uaW5nIHByb3Bvc2FsIHRha2luZyBhbGwKICAgIC8vLyBwcmV2aW91cyB2b3RlcyBpbnRvIGFjY291bnQuCiAgICBmdW5jdGlvbiB3aW5uaW5nUHJvcG9zYWwoKSBwdWJsaWMgdmlldwogICAgICAgICAgICByZXR1cm5zICh1aW50IHdpbm5pbmdQcm9wb3NhbF8pCiAgICB7CiAgICAgICAgdWludCB3aW5uaW5nVm90ZUNvdW50ID0gMDsKICAgICAgICBmb3IgKHVpbnQgcCA9IDA7IHAgPCBwcm9wb3NhbHMubGVuZ3RoOyBwKyspIHsKICAgICAgICAgICAgaWYgKHByb3Bvc2Fsc1twXS52b3RlQ291bnQgPiB3aW5uaW5nVm90ZUNvdW50KSB7CiAgICAgICAgICAgICAgICB3aW5uaW5nVm90ZUNvdW50ID0gcHJvcG9zYWxzW3BdLnZvdGVDb3VudDsKICAgICAgICAgICAgICAgIHdpbm5pbmdQcm9wb3NhbF8gPSBwOwogICAgICAgICAgICB9CiAgICAgICAgfQogICAgfQoKICAgIC8vIENhbGxzIHdpbm5pbmdQcm9wb3NhbCgpIGZ1bmN0aW9uIHRvIGdldCB0aGUgaW5kZXgKICAgIC8vIG9mIHRoZSB3aW5uZXIgY29udGFpbmVkIGluIHRoZSBwcm9wb3NhbHMgYXJyYXkgYW5kIHRoZW4KICAgIC8vIHJldHVybnMgdGhlIG5hbWUgb2YgdGhlIHdpbm5lcgogICAgZnVuY3Rpb24gd2lubmVyTmFtZSgpIHB1YmxpYyB2aWV3CiAgICAgICAgICAgIHJldHVybnMgKGJ5dGVzMzIgd2lubmVyTmFtZV8pCiAgICB7CiAgICAgICAgd2lubmVyTmFtZV8gPSBwcm9wb3NhbHNbd2lubmluZ1Byb3Bvc2FsKCldLm5hbWU7CiAgICB9Cn0=) works for me in Firefox.

So it will work at least for some people. The question would if it's enough of a problem to start looking for an alternative solution.

axic replied:
Can you still add a warning here in case it is longer than 2000 characters?

cameel replied:
> Can do that easily from the user side.

Do you mean that just using `POST` from the client side should work? Does not seem to be the case. I checked now and looks like it's behind Cloudfront and `POST` requests do not even get through to it:

```bash
curl --request POST https://remix.ethereum.org/ --data "code=Y29udHJhY3QK&version=soljson-v0.8.4+commit.c7e474f2.js" --header "Host: remix.ethereum.org"
```
```
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<HTML><HEAD><META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
<TITLE>ERROR: The request could not be satisfied</TITLE>
</HEAD><BODY>
<H1>403 ERROR</H1>
<H2>The request could not be satisfied.</H2>
<HR noshade size="1px">
This distribution is not configured to allow the HTTP request method that was used for this request. The distribution supports only cachable requests.
We can't connect to the server for this app or website at this time. There might be too much traffic or a configuration error. Try again later, or contact the app or website owner.
<BR clear="all">
If you provide content to customers through CloudFront, you can find steps to troubleshoot and help prevent this error by reviewing the CloudFront documentation.
<BR clear="all">
<HR noshade size="1px">
<PRE>Generated by cloudfront (CloudFront)
Request ID: wNDoDUYB6ed6IWHQX9SVQJsn0hB715V4DKOvkdMTJIJ5pfeL2y2j7g==
</PRE>
<ADDRESS>
</ADDRESS>
</BODY></HTML>
```
Only `GET` requests work:
```
curl --request GET https://remix.ethereum.org --header "Host: remix.ethereum.org"
```

cameel replied:
> At the minimum you can output a warning on `make docs` if any snippet has an URL longer than 2000 bytes. We could then see if anything is affected.

ok, warning added. With the limit at 2000 bytes there are 32 snippets that do not pass it. With 4000 bytes it's 18 snippets. With 6000 bytes it's only 3 snippets. At 8000 bytes all the snippets pass.

axic replied:
> > Can do that easily from the user side.
>
> Do you mean that just using POST from the client side should work? 

No, actually I meant to say `can't`, just made a typo.

cameel replied:
> Just ask remix to support lib compression and/or ascii85 (with a modified characters removing % and &) in the payload :)

That's probably not a serious comment but with all snippets being below 8k bytes maybe we could fit in 2k chars with good compression. Code is very squishy.

The snippet from `voting.rst` (one of the 3 longest) has 5415 bytes and just zipping it at default settings reduces it to 2055 bytes. The base64-encoded URL is a little bigger but not by all that much (7253 bytes for that snippet).

cameel replied:
> No, actually I meant to say can't, just made a typo.

Ah, ok :)

Anyway, given that this is a new feature and nobody is really relying on it yet - maybe it would be ok to go with a limit of 8k chars and call it a day? From the post you linked it looks like both Cloudfront and most of the popular browsers (except for old IE) should handle at least that much. We can always start looking for an alternative solution once we get many reports that these links do not work.

cameel commented:
I just checked epub as requested in the issue and unfortunately this extension does affect the output in a bad way:

![solidity-snippet-epub](https://user-images.githubusercontent.com/137030/127042366-ac2460c7-7e5a-4228-8f49-9f39e9dee744.png)

I tried to disable it for this format but looks like epub still uses the HTML builder so I need to find a better way to filter it out. I'll go back to this later, for now I'm switching the PR to draft.

chriseth commented:
This is really great! It might be a bit difficult to understand what the purpose of the logo is, so maybe some text (in addition) would be better? ("open in remix")

cameel commented:
> It might be a bit difficult to understand what the purpose of the logo is, so maybe some text (in addition) would be better? ("open in remix")

@chriseth I could add a small `REMIX` or `OPEN` below but I think that the whole `open in remix` would be too long. I think it would be better as a tooltip (via `title` attribute) but unfortunately I do not know yet how to add it via Sphinx API.

Maybe it would be better to just design or find a better icon? I asked @franzihei about it today and she said she'll try to come up with something.

chriseth commented:
Yeah, a new icon also sounds good - it should maybe contain an arrow or something...

franzihei commented:
I tried out some stuff and I think the Remix logo might be a bit too much. I tend to just using an icon that everybody understands, like an arrow or a similar "open in"/share icon. 

I think we could even get rid of the text then. Here are some ideas:

![image](https://user-images.githubusercontent.com/41991517/127191427-d3a4f2af-fca6-4d1a-9db8-abff7bc1617e.png)
![image](https://user-images.githubusercontent.com/41991517/127191475-1173f0e0-94d5-4f67-9d34-25e714b889c7.png)

axic commented:
<img width="232" src="https://user-images.githubusercontent.com/20340/127191875-d8cf850e-9f9e-441f-97aa-a56c502088df.png">

I think this could work the best, given it is the smallest, so it won't be distorted as much when scaled down.

cameel commented on docs/ext/remix_code_links.py:
```diff
@@ -0,0 +1,56 @@
+    # NOTE: base64 encoded data may contain +, = and / characters. Remix seems to handle them just
+    # fine without any escaping.
+    base64_encoded_source = base64.b64encode(source_code.encode('utf-8')).decode('ascii')
+    return f"https://remix.ethereum.org/?code={base64_encoded_source}"
```
@axic
> Does remix has a version parameter still, or will it just find it out based on the pragma?

Too many things in a single thread so answering his in a separate one.

It does. The problem is that it looks like this: `version=soljson-v0.8.4+commit.c7e474f2.js`. So if we wanted to include version in the link then the extension would have fetch version list at build time. It would also have to parse the version pragma and choose some matching version (the newest one?).

Do you think it's worth the complication? In the latest docs most snippets work with the latest compiler so the default version will be fine most of the time. On the other hand in older docs that's not necessarily the case so maybe it's worth it just for that.

An alternative solution would be to add an option to code blocks that lets you set compiler version manually. We would then have to manually add it to code blocks though and update with each new version (but that could be automated that with a script that we would run on the docs after release and then commit the modified docs).

cameel commented:
Here's a preview. It does not look good due to non-transparent background but at least you can see the final size:

![open-in-remix-icon1-highlighted-block](https://user-images.githubusercontent.com/137030/127209146-cf84c12d-8466-434d-a7a0-e9b05958925e.png)
![open-in-remix-icon2-highlighted-block](https://user-images.githubusercontent.com/137030/127209148-7163d0bf-cf14-4f82-a410-6fe41bdaffc3.png)
![open-in-remix-icon3-highlighted-block](https://user-images.githubusercontent.com/137030/127209149-805c4b03-fa12-422c-866b-eee8a3eefc78.png)

The one without the box is the most minimalistic and looks best to me. Maybe its text should be bigger though.
Actually, if we agree that we do want text there, we could just add it in CSS, it does not have to be a part of the icon unless we want a fancy font.

> We can't do a hover or can we? If yes, it would be great if we could show "open in Remix" when hovering over the icon and then we didn't need the text.

It's possible in HTML but I don't know how to do it through Sphinx API. I have already spent some time trying to figure it out but I eventually gave up because it was starting to take an unreasonable amount of time compared to the rest of the PR. I could try harder if you think it's really essential though.

franzihei commented:
> I could try harder if you think it's really essential though.

No it's not necessary imo.

Thank you for the previews in the respective sizes, I think I prefer the 3rd option with the simple arrow. What do you think?

chriseth commented:
I also like the third one best. Maybe we could alter the font sizes of `open in` and `Remix` so that `Remix` is larger and `open in` a bit smaller? Currently, "Remix~ is just barely readable.

cameel commented:
As I said on the call, this is waiting only for a better quality image for the icon, with transparent background (looks like the consensus is that we want the third one) and for a decision regarding https://github.com/ethereum/solidity/pull/11697#discussion_r67683671. I think that the URL length limit of 8k chars would be reasonable and all our snippets are under that. With the 2k limit suggested by @axic we have 32 snippets that don't pass (which is why there are CI failures). If we keep the lower limit, we need to decide what to do about these snippets.

franzihei commented:
The icon can be downloaded here: https://thenounproject.com/term/export/37875/.

We can also use Fontawesome if that's easier, here is a similar icon: https://fontawesome.com/v5.15/icons/share?style=solid

cameel commented:
ok, that solves the icon problem. How do we deal with attribution? The [CC license](https://creativecommons.org/licenses/by/3.0/us/legalcode) says that we have to credit the author if we distribute it.

franzihei commented:
What about just using the fontawesome one?

cameel commented:
OK, looks like that one does not require attribution:

https://fontawesome.com/v4.7/license/

> Attribution is no longer required as of Font Awesome 3.0 but is much appreciated: "Font Awesome by Dave Gandy - http://fontawesome.io".

cameel committed:
Include language and compiler version in Remix code links

cameel committed:
Add an arrow icon for Remix link, with proper attribution

cameel commented:
I have added the new icon. Here's how it looks like:

![open-in-remix-icon-font-awesome](https://user-images.githubusercontent.com/137030/133988097-5c4bea0e-bb5a-44c7-9859-d1b68a6a9a1b.png)

Looks like font awesome requires attribution too after all (there's a pop up saying that when you click the download button). Fortunately it looks like we do have a place for this stuff: [`cmake/templates/license.h.in`](https://github.com/ethereum/solidity/blob/develop/cmake/templates/license.h.in). I added a short note with a link there.

I have also added version and language parameters to the URLs. Looks like the implementation of these features is not yet done in Remix (https://github.com/ethereum/remix-project/issues/1489, https://github.com/ethereum/remix-project/issues/1490) but currently it simply ignores them so we might want to add them anyway.

Remaining problems:
- The link overlaps the text snippet if the snippet is wide enough to require scrolling. I don't know how to deal with that other than moving the link outside of the box.

    ![open-in-remix-icon-font-awesome-overlapping-code](https://user-images.githubusercontent.com/137030/133988101-db12af8a-3ac8-4c6b-842e-81820bf2f128.png)
- We still need a decision about the length limit (https://github.com/ethereum/solidity/pull/11697#discussion_r67683671).
- In non-release builds the version parameter will contain a non-existent version (e.g. 0.8.8 right now). Not sure if it's better to leave it like this or to only include it in release builds.

chriseth commented on cmake/templates/license.h.in:
```diff
@@ -112,6 +112,12 @@ evmc:
    See the License for the specific language governing permissions and
    limitations under the License.
 
+docs/_static/img/share-solid.svg
```
This is only added to the binary - maybe we should also just add a section to the documentation?

cameel replied:
You mean for all licenses or just for this? Seems like an overkill if it would be just for this one icon.

chriseth replied:
I mean the icon is not present in the binary, why should he binary contain a copyright notice about the icon?

cameel replied:
Ah, I get what you mean now. OK, I can create a new page for this but this will be the only thing there :)

cameel replied:
I have removed this from `license.h.in` and added a new documentation page instead. Called "Credits and attribution" with a section about website icons, meant to be similar to https://creativecommons.org/website-icons. With a link to the page in the footer, just like they do it.

franzihei commented:
> The link overlaps the text snippet if the snippet is wide enough to require scrolling. I don't know how to deal with that other than moving the link outside of the box.

We could also add a background to the image/ make it look like an overlay? But then this might make the code behind it not visible anymore, I'm not sure how far it is possible to scroll there.

cameel commented:
> I'm not sure how far it is possible to scroll there.

![code-block-scrolled-far-right](https://user-images.githubusercontent.com/137030/134395814-571c472b-8f8e-4cad-becf-e3e4023de879.png)

> But then this might make the code behind it not visible anymore,

Yeah, that's the problem. It covers the end of the snippet.

If you want to keep it in the box, we'd need some magic, but none of it seems too great to me:
- Making it transparent on hover: might be weird if you want to click it and then it almost disappears when you hover :)
- Putting it inside the `<pre>` block so that it scrolls with it: might be hard with CSS alone. Also it would add scrollbars in snippets that just barely fit without scrollbars currently.
- Moving the snippet a few lines down when you hover on the link: jumping snippets are not nice.
- Making it appear only when you hover over the code block: this might be the least intrusive option (the "copy text" icons in code blocks on github work that way) but it will make these links harder to discover if you are on mobile and do not have a mouse.

Maybe we should just put it on the right margin of the page? There's some space there, especially if we put text below the icon. Not sure if it will look good though.

franzihei commented:
From the options outlined this one seems best for me:

> Making it appear only when you hover over the code block: this might be the least intrusive option (the "copy text" icons in code blocks on github work that way) but it will make these links harder to discover if you are on mobile and do not have a mouse.

I don't think people without a mouse will try to do stuff in Remix on their phone, will they?

> Maybe we should just put it on the right margin of the page? There's some space there, especially if we put text below the icon. Not sure if it will look good though.

Where do you mean exactly? Like this?

![image](https://user-images.githubusercontent.com/41991517/134522727-3a183b18-accf-4c1e-a180-852888bfc6bb.png)

If there was a "hover over" option, we could also just show the arrow and show "open in Remix" when hovering over with the mouse.

Then we could consider putting the arrow on the right like this:

![image](https://user-images.githubusercontent.com/41991517/134522954-d5b7d044-4633-4b7d-9855-cf9039348430.png)

or, if there is enough space on the top like this:

![image](https://user-images.githubusercontent.com/41991517/134523028-ce502e82-5db7-48e8-9de1-7971468ce76a.png)

It all doesn't look too perfect, but I guess there aren't many other options available. :/

cameel commented:
> I don't think people without a mouse will try to do stuff in Remix on their phone, will they?

I think we'd be surprised :) At least Remix does seem to work on my phone just fine so wanting to click such a link and compile is not out of the question. And people these days do a lot more stuff on mobile that I'd consider sensible or convenient.

But if you think it's the best option, I can do that anyway.

> Where do you mean exactly? Like this?

Yes. Exactly. Does that look ok? To me it actually looks good in your screenshot.

> If there was a "hover over" option, we could also just show the arrow and show "open in Remix" when hovering over with the mouse.

It's technically possible and there should be a way to do this in Sphinx one way or another. I'd just need to do some digging and experiment because docs on this are very sparse.

Would you prefer this one over the one showing on hover or the one with label?

> or, if there is enough space on the top like this:

This one will be problematic if you don't want an extra empty line to always be there. If the preceding line is long, the icon will overlap

cameel commented:
I just tried to make it appear only on hover and it's actually much harder than I thought because the link is not really inside the block but next to it:
- The link is now just before the code block. Unfortunately CSS does not let you select elements before, only after some element so a selector for hover does not match.
- I can put the link after the block and then hover works but the problem is that I can't position the link relative to the top of the block. Its position is relative to the bottom instead.
- If I put the link inside the block, highlighting breaks completely. Apparently the highlighting plugin sees the element inside `<pre>` as a part of the text to be highlighted and decides it's not a valid Solidity code. It then does not even wrap the block in a `.highlight-solidity` div.

I think we'll have to go with placing it on the side.

franzihei commented:
Ok should we just try the first option with icon and text underneath each other on the right side? I think that could actually look ok!

leonardoalt commented:
What's the conclusion here?

cameel commented:
> What's the conclusion here?

Waiting for some changes from my side and for some review feedback:

1. I need to fiddle with CSS a bit more and try to move the icon to the side.
2. We need to agree how to handle URL length limit (https://github.com/ethereum/solidity/pull/11697#discussion_r676836714, https://github.com/ethereum/solidity/issues/11415#issuecomment-931420156).
3. I need to move the icon attribution to a new page in the docs (https://github.com/ethereum/solidity/pull/11697#pullrequestreview-759551804).

cameel commented:
Hopefully this is now ready:
- Icon is now to the side, with text appearing only on hover.
- Added a "Credits and attribution" page with a link in the footer.
- I have increased the URL length limit to 8000 because I think this is the best compromise.

cameel commented:
Well, looks like 8000 is no longer enough. We do have one Yul snippet longer than that (9577 byte-long URL). It's the ERC20 implementation example from `yul.rst`.

What do we do about it? I'd be inclined to increase the safe length even more (to 10k) because I don't think it's really such a big issue - at worst user can just copy and paste it manually.

A more robust solution would be to have each of these longer snippets served as a separate file and use [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude) directive to also keep it on the page as it is now. Then we could give Remix the URL as suggesed in https://github.com/ethereum/solidity/issues/11415#issuecomment-931424265. This requires changes in the extension to be able to handle such blocks specially and maybe some extra changes to ensure these literalincludes are actually served as separate files. I'm not sure it's really worth the effort. If we want this, I'd prefer to get this PR finally merged and work on it separately as an additional feature.

franzihei reviewed (commented):
I like the hover-over feature and I think it looks well enough design-wise. Of course it's a bit tight on the side, but I think it's the best we can do for now and it will add a real benefit for the users, so that's great! I don't have opinions on the open question re safe length, so once that's resolved this looks good to me! :)

chriseth commented:
build fails.

leonardoalt commented:
I think it's fine to use 10k or more to be safe, then `b_docs` should pass.

cameel commented:
> build fails.

Yeah, it's because of the link length limit (it's a warning and we have "treat warnings as errors enabled").

I have increased the limit to 10k.

Karlitos88 reviewed (commented):
6adfe88211ecbd3f3cbc7128a3f13bedd2e269d8
